cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -294,10 +452,22 @@ class InputState {
   // Index of the time col
   col_index_t time_col_index_;
   // Index of the key col
-  col_index_t key_col_index_;
+  vec_col_index_t key_col_index_;
+  // Type id of the time column
+  Type::type time_type_id_;
+  // Type id of the key column
+  std::vector<Type::type> key_type_id_;
+  // Hasher for key elements
+  mutable KeyHasher* key_hasher_;
+  // True if hashing is mandatory
+  bool must_hash_;
+  // True if null by-key values are expected
+  bool nullable_by_key_;
```

Comment:
Why do we need a bool to control this? (instead of always handling null key?)

Reply:
For performance. There are two code-paths for computing a by-key value:
1. A slower path using hashing, which works for any kind of by-key (single or multi, of any type)
2. A faster path using [norm-value](https://github.com/apache/arrow/pull/13880/files#diff-5789a42aebc3bd0f5a6db687c78ab0c3eef3e316982f99fe9e0ea21fabed354cR67), which works only for a single-primitive-types by-key

We want to use the faster path whenever possible. [This post](https://github.com/apache/arrow/pull/13880#issuecomment-1225741731) explains an alternative that would avoid this flag at the expense of runtime complexity.

Reply:
As implemented now, the faster path just applies `norm_value` to an integral value, while assuming it is not null, so it cannot work for a nullable column. It's possible to implement the fast path differently, such that the null bit is accounted for, by normalizing to a wider type (e.g., a 32-bit nullable value would be normalized to a 64-bit value); this would require special handling for 64-bit values (e.g., normalizing to a yet wider value, or resorting to hashing) and I opted to go with the current implementation. Let me know if you feel strongly about avoiding a nullable option being exposed to the user.

Reply:
I think we do want to avoid to having to specify this by the user - ideally the engine knows if a column is nullable and utilize the fast path without user specifying any flags.

Reply:
I'll look into this, then.

>ideally the engine knows if a column is nullable and utilize the fast path without user specifying any flags.

Unfortunately, The engine doesn't know this. In fact, whether a column is nullable cannot be known upfront; even after an input node has generated a million exec-batches where the column is non-null the next exec-batch could have this column nullable. The only way to deal with this is dynamically, by rehashing the memo-store, which means replacing the fast-path keys with slow-path keys for all entries. This replacement's cost is incurred at most once in the life of an `AsofJoinNode`, so it's not too bad from a performance perspective; it's just more complex logic around hashing.

Reply:
I am leaning towards keeping the logic simple for now and just use the hash path. I haven't seen in other API that the user needs to specific this join option to enable fast path. I think a better/long term solution is for the engine to figure out when the fast path can be used and we can leave it out for now.

Reply:
There are also some benchmark utils for asof join: https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/exec/asof_join_benchmark.cc

Maybe can use those to see how much the perf diff is

Reply:
According to [benchmark results](https://github.com/apache/arrow/pull/13880#issuecomment-1234248082), the latest code in this PR is ~15% slower than its baseline. Since the benchmark only covers single-key, namely the fast-path, then clearly there is already a performance cost without the hashing. Of course, we can expect hashing to have a yet higher performance cost.

I'm fine with getting rid of the `nullable_by_key` option. If it were up to me, I'd vote for spending a bit more time on performance tuning thereafter.

Reply:
> I'm fine with getting rid of the `nullable_by_key` option.

[Fixed](https://github.com/apache/arrow/pull/13880#issuecomment-1235468859).